File: ios/FamilyConnect/Core/ReconnectBackoff.swift
//
//  ReconnectBackoff.swift
//  FamilyConnect
//
//  Full-jitter exponential backoff for the WebSocket reconnect loop
//  (the AWS-blessed variant: delay = random(0, min(cap, base·2ⁿ))).
//
//  Full jitter — rather than the deterministic delay Geo's one-shot HTTP
//  retry uses — because reconnects here ARE a thundering-herd problem: a
//  family's devices all lose the same server at the same moment (server
//  restart, router blip) and would otherwise stampede back in lockstep.
//
//  A pure struct with an injectable random source so tests can pin the
//  exact ceiling sequence (1, 2, 4, … 30, 30) without statistics.
//

import Foundation

nonisolated struct ReconnectBackoff: Sendable {
    /// First ceiling, in seconds.
    let base: TimeInterval
    /// Ceilings never exceed this, in seconds.
    let cap: TimeInterval

    /// Injectable random source. Defaults to the system RNG; tests inject
    /// e.g. `{ range in range.upperBound }` to observe the raw ceilings.
    private let random: @Sendable (ClosedRange<Double>) -> Double

    /// Consecutive failures since the last `reset()`.
    private(set) var attempt: Int = 0

    init(
        base: TimeInterval = 1,
        cap: TimeInterval = 30,
        random: @escaping @Sendable (ClosedRange<Double>) -> Double = { Double.random(in: $0) }
    ) {
        self.base = base
        self.cap = cap
        self.random = random
    }

    /// The delay before the next reconnect attempt, advancing the attempt
    /// counter. Uniform in [0, min(cap, base·2^attempt)].
    mutating func nextDelay() -> TimeInterval {
        let ceiling = min(cap, base * pow(2, Double(attempt)))
        // Saturate rather than overflow if a socket flaps for days.
        if attempt < 62 { attempt += 1 }
        return random(0...ceiling)
    }

    /// Call on a successful connection so the next drop starts cheap again.
    mutating func reset() {
        attempt = 0
    }
}

File: win/src/FamilyConnect.Core/Protocol/ReconnectBackoff.cs
namespace FamilyConnect.Core.Protocol;

/// <summary>
/// Full-jitter exponential backoff — <c>delay = random(0, min(cap, base·2ⁿ))</c> — for the socket's
/// reconnect loop and, borrowed, for the send outbox's own per-message schedule.
/// </summary>
/// <remarks>
/// <para>
/// FULL jitter rather than a deterministic delay, because reconnects here are a thundering-herd
/// problem: a family's devices all lose the same server at the same moment (a restart, a router
/// blip) and would otherwise stampede back in lockstep.
/// </para>
/// <para>
/// The random source is injectable so a test can pin the exact ceiling sequence — 1, 2, 4, … 30,
/// 30 — without statistics. Apple counterpart: <c>ReconnectBackoff</c>.
/// </para>
/// </remarks>
public sealed class ReconnectBackoff(
    double baseSeconds = 1.0,
    double capSeconds = 30.0,
    Func<double, double>? random = null)
{
    private readonly Func<double, double> random = random ?? (ceiling => Random.Shared.NextDouble() * ceiling);

    /// <summary>Consecutive failures since the last <see cref="Reset"/>.</summary>
    public int Attempt { get; private set; }

    /// <summary>
    /// The delay before the next attempt, advancing the counter. Uniform in
    /// <c>[0, min(cap, base·2^attempt)]</c>.
    /// </summary>
    public TimeSpan NextDelay()
    {
        var ceiling = Ceiling(Attempt);
        // Saturate rather than overflow if a socket flaps for days.
        if (Attempt < 62)
        {
            Attempt++;
        }
        return TimeSpan.FromSeconds(random(ceiling));
    }

    /// <summary>Call on a connection that lasted, so the next drop starts cheap again.</summary>
    public void Reset() => Attempt = 0;

    /// <summary>
    /// Position the counter, for a caller whose tally lives somewhere else — the outbox keeps its
    /// own per-message count on the row so a retry schedule survives a relaunch, then borrows this
    /// shape to turn it into a delay.
    /// </summary>
    public void AdvanceTo(int attempt) => Attempt = Math.Clamp(attempt, 0, 62);

    private double Ceiling(int attempt) => Math.Min(capSeconds, baseSeconds * Math.Pow(2, attempt));

    /// <summary>
    /// Whether a finished connection earned its reset. A COMPLETED HANDSHAKE IS NOT ENOUGH, and
    /// that distinction is the whole point: a proxy can accept the upgrade and drop the connection
    /// immediately, and so can this server when a send queue overflows. Forgiving the ceiling at
    /// handshake made the ceiling never climb — the socket reconnected about twice a second
    /// indefinitely, each cycle firing a full resync. Judging by DURATION tells the two apart
    /// without asking the endpoint anything.
    /// </summary>
    public static bool EarnsReset(DateTimeOffset? connectedAt, DateTimeOffset now, TimeSpan durableAfter) =>
        connectedAt is { } since && now - since >= durableAfter;
}
